### DISCUSSION TITLE
Don't understand how reactive works

### DISCUSSION THREAD

**Discussion by sergeymild** [mobxjs/mobx#2866] [General]
Hello i have a code
```
constComponent: React.FC = observer(() => {
    cont store = useContext(MobxStore)
    return (<FlatList data={store.items} renderItem={({item, index}) => <Text>{index}</Text>}/>)
})

class MobxStore {
  constructor() {
    makeAutoObservable(this)
  }

  @observable
  items: Array<number> = []

  removeItem = (index: number) => {
    runInAction(() => this.items.splice(index, 1))
  }
}
```

If i will call `store.removeItem(1)`, In that case the FlatList will not be updated.

But if make a something like that:
```
constComponent: React.FC = observer(() => {
    cont store = useContext(MobxStore)
    const items = store.items
    return (<FlatList data={items} renderItem={({item, index}) => <Text>{index}</Text>}/>)
})

// or if in removeItem method i will make
this.items = this.items.filter((a, i) => i !== index)
```

FlatList will be updated.
What i did wrong? Thanks.

**Comment by mweststrate** [comment]
https://mobx.js.org/react-integration.html#callback-components-might-require-observer

**Reply by sergeymild** [reply]
yes but here i don't need observer for each item, in that case i only need observe arrays of items and delete it from list after i have called removeItem. But flatList still render all items even if i have called removeItem

**Reply by sergeymild** [reply]
My problem is that whole FlatList don't rerenders after i call removeItem

**Reply by mweststrate** [reply]
I don't know what FlatList is, so can't tell when it will render or not. For your other question, please study the linked page first, the answer is quite literally in there.

**Comment by mweststrate** [comment]
Sorry, section above it.

**Comment by urugator** [comment]
https://github.com/mobxjs/mobx-react/issues/484#issuecomment-391685354
